### DISCUSSION TITLE
Trying to install OpenSSL on Raspberry Pi 4 but keep getting an error when trying to execute make

### DISCUSSION THREAD

**Discussion by r8drfn89** [openssl/openssl#21599] [Q&A] [answered]
I followed the commands to try and get the newest version on my raspberry pi4. I am trying to install cmake but it wont configure without an OpenSSL package. So here I am trying to download this but run into an error after executing the make command. I am using 3.1.1 from the OpenSSL main websites downloads page. 
 
$ wget https://www.openssl.org/source/openssl-3.1.1.tar.gz
$ tar -zxvf openssl-3.1.1.tar.gz
$ cd openssl-3.1.1
$ sudo apt install build-essential
$ ./Configure
$ make

After the make it gives me an error1 for makefile:12877 and error 2 for makefile:3256

These are the pictures for the `perl configdata.pm --dump`

![0FD0DEB5-6892-4C2A-8832-4293EF4F081F](https://github.com/openssl/openssl/assets/137663045/bfbdb62a-4431-432c-bdea-dc1c35f73c01)
![A1F55E75-E8AC-4D9C-8ACA-443776DC205D](https://github.com/openssl/openssl/assets/137663045/249355cd-97f7-4a3c-9a61-fbfbc62a30a0)
![E54794D4-6603-4D97-920F-36B052A9A6A3](https://github.com/openssl/openssl/assets/137663045/73d1598b-df96-4716-83d2-76841693cc54)
![2408AC6F-F38E-49E4-AEB9-103B7EB753D7](https://github.com/openssl/openssl/assets/137663045/578d5103-ae14-436b-ba36-444401b527d8)
![1B0A261F-B2CE-4C0C-B705-F82C3C677F6C](https://github.com/openssl/openssl/assets/137663045/feffbf60-acf2-45b5-b0b6-ada077a8f4e7)

**Comment by tom-cosgrove-arm** [comment]
It would be useful to see the output from `Configure` itself, as well as the exact operating system and version, and whether it's 32-bit or 64-bit. (The outputs of `cat /etc/os-release` and `uname -a` would help here.)

It looks like OpenSSL has configured for 64-bit (`asm_arch => "aarch64"`) but you appear to be running on a 32-bit system (Perl is for `arm-linux-gnueabihf-thread-multi-64int`). The information requested might help us figure out why that is happening.

A workaround could be to configure explicitly for 32-bit:

`./Configure linux-armv4`

**Comment by r8drfn89** [comment]
it is 32-bit not sure why It keeps trying to configure for 64 bit.

**Comment by r8drfn89** [comment]
![25D38F22-C6E8-4CA9-B1F1-350F3121CA5C](https://github.com/openssl/openssl/assets/137663045/0ca44c01-df48-4963-83c2-0eaba90d419f)

this is the output from the configure. I am unable to scroll up any further. 
![55B76175-665A-4EC1-ABCD-434870F6E441](https://github.com/openssl/openssl/assets/137663045/f32adecb-2ddd-4b8b-85e3-5125901242fd)

**Comment by tom-cosgrove-arm** [comment, marked as answer]
Ah, `uname` is saying we have `aarch64`, so a 64-bit kernel, but `LONG_BIT` is giving 32, hence a 32-bit OS.

I don't know Raspbian well enough to know why it's reporting 64-bit from `uname` but 32-bit from `LONG_BIT`, some weird combination of 32-bit userland on 64-bit kernel, maybe?

Whatever, `./Configure linux-armv4` is your best bet.

Note that `Configure` output should look like either

```
pi32 $ ./Configure
Configuring OpenSSL version 3.1.1 for target linux-armv4
Using os-specific seed configuration
Created configdata.pm
Running configdata.pm
Created Makefile.in
Created Makefile
Created include/openssl/configuration.h

**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL.md file first)      ***
***                                                                ***
**********************************************************************
```

or

```
pi64 $ ./Configure
Configuring OpenSSL version 3.1.1 for target linux-aarch64
Using os-specific seed configuration
Created configdata.pm
Running configdata.pm
Created Makefile.in
Created Makefile
Created include/openssl/configuration.h

**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL.md file first)      ***
***                                                                ***
**********************************************************************
```

**Comment by r8drfn89** [comment]
![824016AB-D391-460C-A560-E0E79CA1BD15](https://github.com/openssl/openssl/assets/137663045/49b056d3-6575-4056-806d-aeb458c8a8e6)
This was the photo I meant to upload

**Comment by tom-cosgrove-arm** [comment]
So is the `make` after `./Configure linux-armv4` still giving errors?

Note: you might need to `make clean` before `make` since you've built with a different configuration previously.

**Comment by r8drfn89** [comment]
The make is still working. But it looks promising, It has been 5 min now. Not sure how long it should take? The previous times it would give the error after 1 min and stop.

**Comment by r8drfn89** [comment]
The make was successful and now the install is going.

**Comment by r8drfn89** [comment]
The install finished but I get this. 
![1128DE0D-297A-4D93-97BB-B6A9C65B9677](https://github.com/openssl/openssl/assets/137663045/827eb524-9124-4cc9-bd34-8af2b6b8aefc)

**Comment by tom-cosgrove-arm** [comment]
So you now have two versions of OpenSSL installed - the one that came with your operating system (which you shouldn't remove, since the API has changed between v1.1.1 and v3.x, and other system components may stop working if the 1.1.1 version goes away), and the one you have installed in `/usr/local`.

As to how to get CMake to use it, that's a question for a CMake list...

**Comment by beldmit** [comment]
You ran not the freshly built openssl but the system one. I presume the new should be in /usr/local/bin/openssl

**Comment by r8drfn89** [comment]
I understand now.  I was downloading this because I got this error trying to install CMake. 

![58AE07C2-4A61-4AE1-A9D1-42E2D9CB2D29](https://github.com/openssl/openssl/assets/137663045/2a6cd23d-bfd3-4572-a287-2126f8a93541)

**Comment by r8drfn89** [comment]
@tom-cosgrove-arm Thank you for your help. This **A workaround could be to configure explicitly for 32-bit:

./Configure linux-armv4**  fixed the error I was getting for the OpenSSL and now the make for the CMake is halfway done.
